### PR TITLE
Add liftA2 and liftA3

### PR DESCRIPTION
**Pull request by TheLudd**
The lift function specifically implemented for fantasy land compatible
types. Comes in two versions, for functions of airity 2 and 3.

### CODE REVIEW

**Comment by TheLudd** [comment]
We discussed whether to have functions like these in ramda-fantasy or not. I would prefer to include `lift` here. Partly because I prefer to use these version over the generic one and partly because the elegant way they are implemented. 

I know that `lift` already exists in ramda but these would be my preferences when working with Apply instances.

File: src/lift-a3.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+var R = require('ramda');
+
+module.exports = R.curryN(4, function liftA2(f, a1, a2, a3) {
+  return a1.map(f).ap(a2).ap(a3);
+});
+
+
```
Comment (davidchambers) [inline src/lift-a3.js]: :space_invader:

Comment (TheLudd) [inline src/lift-a3.js]: What does that mean? =)

Comment (TheLudd) [inline src/lift-a3.js]: Empty lines?

Comment (davidchambers) [inline src/lift-a3.js]: Yes. They are `:space_invader:`s. ;)

Comment (TheLudd) [inline src/lift-a3.js]: They are shot down

**Comment by buzzdecafe** [comment]
ramda has variadic lift which should cover these cases. do we need these for external compat?

**Comment by CrossEye** [comment]
I don't know.  Should we have `liftM2`, `liftM3`, as well?  I'm not seeing an advantage.

But I haven't been following the direction of the fantasy project very closely.  Is there a goal to make this a very separate stand-alone implementation of the FantasyLand specification with no dependencies on Ramda?  If that's the case, then I could see where this makes sense.

My original thought with this, long ago, and I don't know if @buzzdecafe was thinking the same way, was that an extension like this was something that could easily be bundled with a version of Ramda for those who wanted these additional features.  I never really thought about it standing on its own.

But we have definitely moved away from the bundling idea, and perhaps even the notion of extensions no longer really makes sense.

**Comment by TheLudd** [comment]
@CrossEye I *do* see this as an addon to ramda. Using these types without ramda doesnt make much sense to me. This is also why I haven been shy to import ramda where needed.

The main difference for me (which I have mentioned before so apologies if I am repeating myself) is that you can do `liftA2(f, a1, a2)` in one go with this function. That is not possible with the current implementation of ramdas lift. You need `lift(2, f)(a1, a2)`. Plus, I actually prefer the specific arity as opposed to passing it in as the first argument.

> I don't know. Should we have liftM2, liftM3, as well? I'm not seeing an advantage.

I don't know what these functions implies actually, that they are for monads? As I see it, the functions I implemented requires the input argumets to be of type `Apply`. Which is what I assume the `A` in `liftA` stands for (I did not come up with the name myself). As a `Monad` is also an `Apply`, no specific lift functions for them will be needed.

**Comment by CrossEye** [comment]
I hadn't caught the difference we had between `lift(2, f)(a1, a2)` and `liftA2(f, a1, a2)`.  I agree that the latter is more Ramda-like.  It's also more closely aligned with the recent goals of updating variadic functions into fixed-arity ones.  I'm still only half-committed to that goal, but if we're going to do it, we're going to have to revisit `lift` in any case.

But two things:  First, I think we need to address this in Ramda itself.  `lift`/`liftN` was a nice experiment, but it looks as though it might not be the best bet and a collection of `lift2`, `lift3`, `lift4` would be better.  Second, the name.  I'm pretty sure the Haskell names are based on Applicatives and Monads, which, at least until recently, didn't actually share the obvious relationship.  I think we simply don't need to care about that, and can deal with them in a more generic way and simply call the function `lift`, with the only suffix being the parameter count.

**Comment by buzzdecafe** [comment]
> deal with them in a more generic way and simply call the function lift, with the only suffix being the parameter count.

in which case the existing abstractions of `lift/liftN` are already there. Problem solved!

**Comment by buzzdecafe** [comment]
ping -- how are we going to proceed here?

**Comment by TheLudd** [comment]
Well, I'd prefer to have these versions in. At least until it is decided if ramdas `lift` goes through any change or not. And for that reason I'd also prefer to keep the names with the A at the end to clarify which is which.

**Comment by buzzdecafe** [comment]
i'm inclined to agree with @CrossEye that the `A` or `M` doesn't add much for us. consider the Haskell sigs:

```hs
liftA2 :: Applicative f => (a -> b -> c) -> f a -> f b -> f c
liftM2 :: Monad m => (a1 -> a2 -> r) -> m a1 -> m a2 -> m r
```

for weakly-typed JS, those amount to the *same thing*. But `lift2, 3, 4` would be fine

**Comment by CrossEye** [comment]
Agreed.  My only question is whether we should just do these in Ramda proper or do it in the fantasy project.  I vote for replacing Ramda's `lift` with `lift2/3/4`.

**Comment by TheLudd** [comment]
The reason for the A as I see it is that the implementation is not a generic lift. It is a lift compatible with fantasy land Apply types.

**Comment by CrossEye** [comment]
Woulld not the identical implementation work for Monads?

I'm also curious about a generic implementation that simply tweaks `liftN`:

    var liftN = _curry2(function liftN(arity, fn) {
      var lifted = curryN(arity, fn);
      var lift = curryN(arity, function() {
         return _reduce(ap, map(lifted, arguments[0]), _slice(arguments, 1));
      });
      return arguments.length > 2 ? lift.apply(this, _slice(arguments, 2)) : lift;
    });

Would that cover everything under consideration?

**Comment by buzzdecafe** [comment]
> return arguments.length > 2 ? lift.apply(this, _slice(arguments, 2)) : lift;

pretty much, AFAICT

**Comment by TheLudd** [comment]
> Woulld not the identical implementation work for Monads?

Yes but the lowest common denominator would be Apply. 

But right. If the generic lift is improved I have less need for this.

**Comment by TheLudd** [comment]
@buzzdecafe Can I take ramda/ramda#1058 as we go with this one? But I change the name to `lift2/lift3`?

**Comment by buzzdecafe** [comment]
yes, exactly. i think `lift_` is a better fit here.

**Comment by TheLudd** [comment]
And we are satsfied with 2 and 3?

**Comment by buzzdecafe** [comment]
that should be enough to get us going. if it turns out we need 4 (or more???) it shouldn't be too much trouble to add it.

**Comment by TheLudd** [comment]
Changes done. I also added them to the index file.

**Pull request merged**